neomodel/util.py

```diff
@@ -193,17 +219,58 @@ def cypher_query(self, query, params=None, handle_unique=True, retry_on_session_
         try:
             # Retrieve the data
             start = time.clock()
-            response = session.run(query, params)
-            results, meta = [list(r.values()) for r in response], response.keys()
+
+            delay = 1
+            while True:
+                last_exception = None
+                cts = time.clock()
+
+                if cts - start > max_retry_seconds:
+                    # keep in mind that if many
+                    # different exceptions are
+                    # thrown, we only show the
+                    # one that occured last
+                    raise last_exception
+
+                try:
+                    response = session.run(query, params)
+                    results, meta = (
+                        [list(r.values()) for r in response],
+                        response.keys(),
+                    )
+                    break
+                except Exception as e:
```

Comment:
Yes, that's right!

I think the neobolt NotALeader exception is a good pick for when the we're now talking to a follower. As a bonus, given how it's named I assume it also will throw an exception when the node is in CANDIDATE status, before the cluster forms.

A good example of (2) is that the app has been directed (say) to node 1, which is the leader, and node 1 dies for some reason. Nodes 2 and 3 re-establish themselves as follower and leader, but while node 1 is physically coming back up it may be at a point where it's not even serving network requests, let alone have cluster info. So if the app tries to connect in that time, it almost certainly will get a connection timeout. When it tries to connect to the old node before it's got the cluster info, that's probably the service unavailable exception. I imagine that neobolt's ConnectionExpired and ServiceUnavailable exceptions will cover those cases.

In the case of ConnectionExpired, it could probably stop trying to connect to that server and connect afresh to the globally configured connection string instead, as that'll probably be a load balancer that will route the traffic to an alive node, re-get routing info and all will be well. What do you think?

Reply:
ah I hadn't thought about the CANDIDATE stage, but I think you're right!

Your example of scenario 2 seems entirely on point, it's consistent with what we see happen every once in a while. 

I agree: connect afresh to the globally configure connection string and it should resolve. I see one possible risk: there might be a small window where the load balancer has not picked up the unavailability of node 1 and so there is a (very) small chance that connecting afresh might end up at node 1 - I'm not sure of the likelihood of this happening (it also depends on health check interval settings and cluster size), but it might be prudent to ensure some retry logic for the reconnect with a configurable MAX_RETRY_SECONDS (which should then be set by the user to a value greater than the health check interval, with a reasonable default - 10 seconds?).

Reply:
@robertlagrant is there any chance you can try this to see if it resolves the issues?

Reply:
@robertlagrant ping?